### ISSUE TITLE
Specifying procedure name with /proc

### ISSUE THREAD

**Issue by crazykt** [boogie-org/boogie#16]
I want to check only some specific procedures in a boogie file. When I try to use the /proc:abc option Boogie verifies all the procedures which contain the string "abc" in their names.
I am running Boogie on Ubuntu-14.04.
```
mono /path/to/Binaries/Boogie.exe /proc:"abc" /trace /traceTimes test.bpl
```
Above command even verifies procedures with names "abcd" , "function_abc" , etc. How can I verify some specific procedure?

**Comment by delcypher** [comment]
@crazykt Thanks for reporting this. This seemed like a bug to me so I've fixed it in 7f4e6b0fab58bb3028cd0f1734fc97b3feafefdf

**Comment by cpitclaudel** [comment]
Wait, this is a huge change to the semantics of that flag

**Comment by cpitclaudel** [comment]
I doesn't seem like a bug to me; it's a great way to specify which procedure you want to verify; when dealing with procedures generated by other tools like Dafny it makes a lot of sense to just give a substring of the name instead of the full `Module$$Impl$$ProcName`

**Comment by cpitclaudel** [comment]
It's a bug that you can't specify a name exactly, but it was a feature that you could specify a name partially. What about checking Contains first, and reverting to substring if that doesn't match anything?

**Comment by delcypher** [comment]
> I doesn't seem like a bug to me; it's a great way to specify which procedure you want to verify

I disagree, there is nothing anywhere that Boogie should behave the way it used to. The help output just says...

```
/proc:<p>      : limits which procedures to check
```

I and probably others (@zvonimir , I know you use this flag, do you have an opinion on this?) have always assumed based on the help description that this would cause boogie to only verify the named procedure and that you could specify ``/proc:`` multiple times to select a set of procedures. To me that is the intuitive behaviour. The previous behaviour of the flag is not intuitive at all.

If it's really necessary to select multiple procedures with a single command line flag then we should probably add a flag that takes a regex, e.g. ``/procMatchRegex:``

**Comment by delcypher** [comment]
I think other devs ( @RustanLeino @shazqadeer @akashlal  ??) need to weigh in here as @cpitclaudel and I clearly have very different opinions on this.

**Comment by akashlal** [comment]
We should support both options (why not?) and update the Boogie help
message accordingly. I suggest treating the argument of /proc as a RegEx.
By default, this means an exact match. For checking for substrings, one
would explicitly need to give wildcards, e.g., /proc:*blah*

Akash

**Comment by zvonimir** [comment]
While we do not really rely on /proc in SMACK when using Boogie, I still do like Akash's suggestion. That way one would have the flexibility to be as specific or general as they want.

**Comment by delcypher** [comment]
@akashlal @cpitclaudel @zvonimir @afd

A problem with Akash's proposal is that ``$`` which is allowed in a procedure name also has a particular meaning (end of line) in regular expressions. This symbol is used by GPUVerify tool.

I think having both options is fine but that they should work like so

* ``/proc:NAME`` - Only try to verify a procedure whose name matches ``NAME`` exactly. This option can specified multiple times to have multiple procedures verified. E.g. ``/proc:foo /proc:bar`` would cause Boogie to try and verify ``foo`` and ``bar`` but no other procedures.

* ``/procRegex:REGEX`` - Only try to verify procedures that match the regular expression ``REGEX``. This option can be specified multiple times to have multiple sets of procedures verified. E.g. ``/proc:^foo.+ /proc:^bar.+`` would cause Boogie to try to verify any procedure whose name begins with ``foo`` or ``bar``.

The old behaviour of ``/proc`` should not be reimplemented because ``/procRegex`` offers a superset of the functionality previously available.

Both options should be properly documented in the help output to make it **very clear** what their behaviours are.

**Comment by zvonimir** [comment]
But isn't /proc subsumed by /procRegex?
For example, what happens if you provide just /procRegex:foo?

**Comment by delcypher** [comment]
But isn't /proc subsumed by /procRegex?

Yes but ``/procRegex`` isn't particularly convenient.

> For example, what happens if you provide just /procRegex:foo?

I need to try implementing this but glancing at C#'s Regex API this would do a match so it would look for an occurrence of ``foo`` anywhere in the procedure name (e.g. it would match ``foo`` and ``myfoo``). To specify exact match you would need to do ``/procRegex:^foo$`` which isn't as convenient as ``/proc:foo``.

**Comment by cpitclaudel** [comment]
Here's a different suggestion: since substring matching seems to have been sufficient until now (it was what was implemented until now), we could support just 4 types of patterns:

* `abc`: Procedures named `abc`
* `*abc`: Procedures ending in `abc`
* `abc*`: Procedures starting in `abc`
* `*abc*`: Procedures containing `abc`

Full regexp support will be annoying to use, because $ signs would be hard to type. We can implement it later if someone really needs it; meanwhile, just adding support for "*" at the beginning or at the end of an expression is easy to implement, is more general than what we already have, does not require an extra option, and is easy to implement.

**Comment by RustanLeino** [comment]
The substring behavior of /proc was both intended and is useful.  Regrettably, the /help flag is not precise about this.  The primary use of Boogie is as an Intermediate Verification Language, so most programs are generated by a machine.  Such machine translation often makes names long, and thus it would be awkward to have to specify the entire name of a procedure on the command line.

I agree the /proc flag is broken, however, since if there are two procedures, one of whose name is a substring of the other's, then there is no way to specify just the procedure with the shorter name.

Btw, in the Spec# front-end to Boogie, there were several command-line options similar to /proc (see http://specsharp.codeplex.com/SourceControl/latest#SscBoogie/Source/BytecodeTranslation/SscBoogieOptions.ssc):  /translate, /trMethod, /trClass, /trFile, /trExclude.  These were specific to how names were generated in the Boogie program.  We could certainly support more than one switch in Boogie.  For example, /proc could continue to filter according to the given substring and a new flag, like /procedure, could filter according to the full name. 

Of the variations mentioned, I like something simple like Clement's proposal the best.  That is, I think that going for a regular expression seems like overkill and introduces, as has been mentioned, all sorts of issues with escaping characters.  So, how about the following:

/proc will do an exact match, but will allow occurrences of * in any position in the search pattern.  * is a wildcard that stands for 0 or more characters.

This would make it easy to get the /proc functionality that I use every day with Dafny, simply by me having to include one or two stars.  It support the full-match functionality requested.  It does not support more general functionality than has been needed to date.  And compared to a full regular expression, it is lightweight, both in terms of the flag name /proc and in terms of the given pattern.

Furthermore, as is currently the case, no /proc flag means "all procedures", where as 1 or more /proc flags says to verify a procedure if it is requested by any one of the /proc flags.

Finally, we should fix the /help message to be more precise. 

  Rustan

**Comment by parno** [comment]
Has anyone looked at this recently?  We also rely on the original semantics, for some of the reasons that Rustan mentioned, and we're currently using Python scripts to work around the recent change.

**Comment by delcypher** [comment]
@parno I've just made a PR that fixes this. Feel free to review.

**Comment by parno** [comment]
Awesome, thank you!  The patches look good to me, and your test suite looks suitably evil.

**Comment by delcypher** [comment]
Fixed by PR #22 which is now merged.
